src/librustc_codegen_ironox/value.rs

```diff
@@ -28,22 +33,52 @@ pub enum Value {
     ConstInt(usize),
     /// The index of an `IronOxStruct` in the `structs` vec from `ModuleIronOx`.
     ConstStruct(usize),
-    /// The parameter of an `IronOxFunction`. This is just a wrapper around a `Type`.
-    Param(Type),
+    /// The parameter of an `IronOxFunction`. This is just a wrapper around a
+    /// `Type`. A parameter is an (index, type) pair, where 'index' is the
+    /// index of the parameter in the list of parameters of the function.
+    Param(usize, Type),
+    /// An instruction: (functiton index, basic block index, instruction index).
+    Instruction(usize, usize, usize),
     /// A placeholder for unimplemented Values. This variant will be removed.
     None,
 }
 
-#[derive(PartialEq, Copy, Clone, Debug)]
+/// An IronOx instruction.
+#[derive(PartialEq, Debug, Eq, Hash)]
 pub enum Instruction {
-    // FIXME: implement
+    /// Call(fn_idx, args). Emit a call to the function found at index `fn_idx`
+    /// in the `functions` vector of the module.
+    Call(usize, Vec<Value>),
+    /// Allocate space on the stack for a variable of a particular type and
+    /// alignment.
+    Alloca(String, Type, Align),
+    /// Cast a value to a type.
+    Cast(Value, Type),
+    /// Add two values and return the result.
+    Add(Value, Value),
+    /// Currently serves as a no-op. This variant will be removed.
     None,
 }
 
-impl Eq for Value {}
-
-impl Hash for Value {
-    fn hash<H: Hasher>(&self, hasher: &mut H) {
-        (self as *const Self).hash(hasher);
+impl Instruction {
+    pub fn val_ty(&self, cx: &CodegenCx, module: &ModuleIronOx) -> Type {
```

Comment:
So this is getting the type that would result from evaluating an instruction?

Reply:
Yep, I'll add a comment so it's clear.